### ISSUE TITLE
content.json: File permissions / Rate limiting

### ISSUE THREAD

**Issue by python273** [HelloZeroNet/ZeroNet#1071]
### Suppose:
1. User can modify their own `data.json`
2. `data.json` contains list of messages (e.g. forum posts)
3. User can add as many posts as they want since they don't have any restrictions
4. ???
5. FLOOD or MODIFYING POSTS AFTERWARDS!11

### File per post + Rate limiting + File permissions:

1. Allow to add one file per `content.json` update
2. Rate limit modification of `content.json` (check via `modified` key)
3. One file contains one post
4. User can only create files, not modify

#### Problems:
1. Many files

### Smarter ways to check modifications of JSON data:

1. Ability to restrict addition of one item to `data.json` list per `data.json` update via `content.json` rules
2. Ability to restrict items modification
3. Rate limit modification of `data.json` or `content.json`

#### Problems:
1. (Possible) DoS attack via big JSON files (each node should check modification of files)

#### Problems:
1. Kinda dangerous (sandbox's required)

**Comment by HelloZeroNet** [comment]
There is rate limit on content.json updates on same file (15sec) and the data.json file size is limited to each user.

I don't think if rate limiting of post inserts is solves any problem. 
What rate do you limit it to? Even 1 post/min could ban normal users and the spammer could easily send over 1000 post / day.
What if the user was offline for some hours and that time he posted some replies? What if you were offline?

I think this problem should be solved by the moderator of the site. Maybe with help of some automated "bot" to watch the updates and ban obviously spamming users.

**Comment by python273** [comment]
Well I think 1000 posts/day is better than 100500 posts/`data.json` update. And the second problem is modifying post after posting

> I think this problem should be solved by the moderator of the site. Maybe with help of some automated "bot" to watch the updates and ban obviously spamming users.

Yeah, but it's centralized solution, so I'm trying to figure out some decentralized solutions

Is there some API to watch file updates?

**Comment by HelloZeroNet** [comment]
I think most of the users would really hate it if they were unable to modify their post after they sent it.
It could be possible to add more limitation on rate on same file updates. eg. 100/hour (configurable by site owner)

**Comment by python273** [comment]
> I think most of the users would really hate it if they were unable to modify their post after they sent it.

Well in some cases it's better not to provide option to edit messages. Now we can modify data even if a site doesn't provide that option (just edit `data.json`, sign, publish)

> It could be possible to add more limitation on rate on same file updates. eg. 100/hour (configurable by site owner)

The problem is that it's possible to add lot of posts at 1 update, so it won't work at all

btw, as I understand we can't rely on `modified` value. Each node just checks that it's bigger that previous `modified` timestamp. And you mentioned about offline. So we need to have some history of file changes (at least hashes) (like blockchain, lol)

**Comment by imachug** [comment]
There is a limit on file size, so flooding is not possible.

**Comment by python273** [comment]
Even if you have 1MB limit, you can still create file that contains thousands of posts. It doesn't solve flooding problem at all

**Comment by mkg20001** [comment]
@python273 That's because the timestamp in the posts can not be verified. So ratelimiting is useless as one could just use an older timestamp as there is no blockchain to verify whether that piece of content really was created at that time. So a limit like 1 post per minute makes no sense.

**Comment by python273** [comment]
Actually timestamp can be trusted in some range https://en.bitcoin.it/wiki/Block_timestamp

**Comment by mkg20001** [comment]
@python273 ZeroNet only uses BIP-32 crypto. NOT bitcoin blocks or anything else related to bitcoin.

**Comment by python273** [comment]
I know that. It's just an example of verifying timestamp

**Comment by mkg20001** [comment]
@python273 But this implies that the blocks reference the parent (creating a chain aka blockchain). This is NOT the case for zeronet content.json's. They do not use timestamps in a way that they are verifiable. The only assumptions that zeronet guarantees for the timestamp is that it's less than 12h in the future and it's positive.

**Comment by python273** [comment]
I know that too. Read comments before leaving a comment
> as I understand we can't rely on `modified` value. Each node just checks that it's bigger that previous modified timestamp. And you mentioned about offline. So we need to have some history of file changes (at least hashes) (like blockchain, lol)

**Comment by mkg20001** [comment]
@python273 Sorry. Just didn't read the whole thing again.

**Comment by python273** [comment]
I know it's not so easy problem to solve and usually it solved with PoW (like in BitMessage). But if you want to make ZeroNet fully decentralized, it should be solved somehow IMHO

**Comment by mkg20001** [comment]
I don't think that this is the only problem required for full decentralization. We are **trusting** hub owners not to change their minds and do stuff like revoking the permission that allows users to post their stuff in the hub.
